firedrake/pointcloud.py

```diff
@@ -0,0 +1,283 @@
+from firedrake.mesh import spatialindex
+from firedrake.dmplex import build_two_sided
+from firedrake.utils import cached_property
+from firedrake.petsc import PETSc
+from firedrake import logging
+from pyop2.mpi import MPI
+from pyop2.datatypes import IntType, ScalarType
+import numpy as np
+
+
+__all__ = ["PointCloud"]
+
+
+def syncPrint(*args, **kwargs):
+    """Perform a PETSc syncPrint operation with given arguments if the logging level is
+    set to at least debug.
+    """
+    if logging.logger.isEnabledFor(logging.DEBUG):
+        PETSc.Sys.syncPrint(*args, **kwargs)
+
+
+def syncFlush(*args, **kwargs):
+    """Perform a PETSc syncFlush operation with given arguments if the logging level is
+    set to at least debug.
+    """
+    if logging.logger.isEnabledFor(logging.DEBUG):
+        PETSc.Sys.syncFlush(*args, **kwargs)
+
+
+class PointCloud(object):
+    """Store points for repeated location in a mesh. Facilitates lookup and evaluation
+    at these point locations.
+    """
+    def __init__(self, mesh, points, *args, **kwargs):
+        """Initialise the PointCloud.
+
+        :arg mesh: A mesh object.
+        :arg points: An array of points.
+        """
+        self.mesh = mesh
+        self.points = np.asarray(points)
+        _, dim = points.shape
+        if dim != mesh.geometric_dimension():
+            raise ValueError("Points must be %d-dimensional, (got %d)" %
+                             (mesh.geometric_dimension(), dim))
+
+        # Build spatial index of processes for point location.
+        self.processes_index = self._build_processes_spatial_index()
+
+    @cached_property
```

Comment:
If this is a cached property, what is the mechanism for blowing the cache if the user moves the coordinate locations on the mesh?

Reply:
Ideally to be safe, the `PointCloud` would register a callback on the mesh such that the cache gets blown every time `mesh.coordinate.dat.data` gets touched.

Reply:
I suppose that means we need to resurrect object versioning. I will add that to my lazy-evaluation purge.